### DISCUSSION TITLE
Set something up so that optimisation/inference can be stopped and restarted

### DISCUSSION THREAD

**Discussion by MichaelClerx** [pints-team/pints#1573] [Ideas/Science]
The ask-and-tell interface (#132) takes care of part of this, but it'd also be good to store the internal state of our optimisers/inferers to disk, perhaps using [pickling](https://docs.python.org/3/library/pickle.html) ?

**Comment by MichaelClerx** [comment]
#TODO: Have a quick look at https://pythonhosted.org/joblib/

**Comment by MichaelClerx** [comment]
@martinjrobins @ben18785 please add Freya to this ticket if she decides to take it up :-)

Rough idea of how to implement it:

- write it as some kind of "deep copy" of every sampler/optimiser's state
- add bits to the ask-and-tell interface that allow the current state to be obtained, and a previous state object to be restored
- add a default implementation using [copy.deepcopy](https://docs.python.org/3/library/copy.html)
- but make sure that copy.deepcopy is not required !
- figure out what to do for wrapper methods i.e. cmaes

**Comment by martinjrobins** [comment]
@MichaelClerx : points 3 and 4 seem contradictory, what do you mean that copy.deepcopy is not required, its part of the standard library isn't it?

**Comment by MichaelClerx** [comment]
Sorry, I meant that the interface shouldn't be written such that implementations require deepcopy. I.E. it should be possible to write a custom implementation instead

**Comment by MichaelClerx** [comment]
Via a different route I've hit on https://joblib.readthedocs.io/en/latest/ again :D
